Improved output of formatNumberReadable

Ruko97 opened the pull request:
Follow up to issue #9601 and pull request #13163.
Based on advice by @cameel in the conversation in issue.
Previously `formatNumberReadable(57896044618658097711785492504343953926634992332820282019728792003956564819968) `would equal to `0x80 * 2 ** 248`, but now it equals to `2 ** 255`.
Also, there was a bug in the test case `test_format_number_readable_signed` in `test/StringUtils.cpp`. `formatNumberReadable(-0xffffffff)` should be equal to `-2 ** 32 + 1` but previously it was tested to be equal to `-2 ** 32 - 1`. The new version of `formatNumberReadable` fixes this problem.

Ruko97 committed:
changed output of formatNumberReadable so that it shows powers of two and one-less-than powers of two in a more compact format

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,122 @@ std::string joinHumanReadablePrefixed
+		"only integer numbers are supported"
+	);
+
+	bool _isNegative = false;
```
Names starting with leading underscore are used for[ parameter variables](https://github.com/ethereum/solidity/blob/develop/CODING_STYLE.md#4-variable-prefixes). Please remove `_` from here and also `_sign` below so it is clear they are local.

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,122 @@ std::string joinHumanReadablePrefixed
-	HexPrefix prefix = HexPrefix::Add;
+	HexPrefix hexprefix = HexPrefix::Add;
+
+	std::string _sign = _isNegative ? "-" : "";
```
Same as before regarding underscore. 
Also, maybe we could declare and define this before so it can be used in the previous `return` values.

Ruko97 replied:
I defined `sign` after the previous return values because the condition for entering the return statement was to have `isNegative` be true. So there wasn't any need to use the value of sign instead of just adding `"-"` before the value

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,122 @@ std::string joinHumanReadablePrefixed
+	);
+
+	bool _isNegative = false;
+	bigint v;
```
```suggestion
	bigint signedValue = isNegative ? (bigint(-1) * _value) : bigint(_value);
	bigint v;
```
What do you think about using `v` as the working var used by the `for` loops? It can then be initialized with `signedValue` inside each `for` instead of being assigned the ternary operator result before the loops.
Also, `signedValue` can be used in the `return` values.

Ruko97 replied:
Yes, I think that would be a great idea.

matheusaaguiar reviewed (changes requested):
Looks good!
Just some code style adjustments and readability suggestions.

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,117 @@ std::string joinHumanReadablePrefixed
+
+	bool isNegative = _value < 0;
+	bigint temp;
+	bigint absoluteValue = isNegative ? (bigint(-1) * _value) : bigint(_value);
```
Maybe a little nitpick, but it seems to me that `absoluteValue` means how far from zero the number is and then it cannot be negative. What about `decimalValue` (or `signedValue` as suggested before)?

Ruko97 replied:
I think signedValue would be better

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,117 @@ std::string joinHumanReadablePrefixed
 	int i = 0;
-	T v = _value;
-	for (; (v & 0xff) == 0; v >>= 8)
+	temp = absoluteValue;
+
+	for (; (temp & 0xff) == 0; temp >>= 8)
```
```suggestion
	for (temp = absoluteValue; (temp & 0xff) == 0; temp >>= 8)
```
This is more compact and has the same readability.

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,117 @@ std::string joinHumanReadablePrefixed
+	// or -N * 2**x + 1 (if negative)
 	i = 0;
-	for (v = _value; (v & 0xff) == 0xff; v >>= 8)
+	temp = absoluteValue;
+
+	for (; (temp & 0xff) == 0xff; temp >>= 8)
```
```suggestion
	for (temp = absoluteValue; (temp & 0xff) == 0xff; temp >>= 8)
```

matheusaaguiar commented:
I think this is in good shape. Please, rebase.

Ruko97 commented:
Pushed a rebased version right now. Please check it.

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
+/// Returns decimal representation for smaller numbers; hex for large numbers.
+/// "Special" numbers, powers-of-two and powers-of-two minus 1, are returned in
+/// formulaic form like 0x01 * 2**24 - 1.
+/// @a T can be any integer variable, will typically be u160, u256 or bigint.
```
```suggestion
/// @a T can be any integer type, will typically be u160, u256 or bigint.
```
Ignore upper case letter `A`, it should be `a`, but github is somehow tagging a user there... 
Just to be clear, suggestion is to change `variable` to `type`

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
+/// @param _value to be formatted
+/// @param _useTruncation if true, internal truncation is also applied,
+/// like  0x5555...{+56 more}...5555
+/// @example formatNumberReadable((u256)0x7ffffff) = "0x08 * 2**24"
```
```suggestion
/// @example formatNumberReadable((u256)0x7ffffff) = "2**27 - 1"
```
Shouldn't this have the same value as in the check at line 119 of `test/libsolutil/StringUtils.cpp` ?

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
-		"only unsigned types or bigint supported"
-	); //bigint does not carry sign bit on shift
+		std::numeric_limits<T>::is_integer,
+		"only integer numbers are supported"
```
```suggestion
		"only integer types are supported"
```

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
 	{
+		std::string suffix = isNegative ? " + 1" : " - 1";
+
 		// 0xFF yields 2**8 - 1 (v is 0 in that case)
```
```suggestion
		// 0xFF yields 2**8 - 1 (temp is 0 in that case)
```
Also, I guess we could remove `(temp is 0 in that case)`.

Ruko97 replied:
It could help other programmers update the code.

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
+
 	if (_useTruncation)
 	{
 		// return as interior-truncated hex.
```
I'd say that the `if` line already makes this clear. I think the comment can be removed.

matheusaaguiar replied:
What about this and the next comment? I think that the code already makes it explicit, so we could remove this line.
```suggestion

```

matheusaaguiar replied:
@Ruko97 , what do you think about this?

Ruko97 replied:
This comment existed in the prior version of the function, I just left it as it is.

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
 			str.substr(len-finalChars, len);
 	}
 
 	// otherwise, show whole value.
```
I think this is redundant too.

matheusaaguiar replied:
```suggestion

```

Ruko97 replied:
Fixed now

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
+	HexPrefix hexprefix = HexPrefix::Add;
 
-	// when multiple trailing zero bytes, format as N * 2**x
+	// when multiple trailing zero bytes, format as N * 2**x (if positive)
+	// or -N * 2**x (if negative)
```
```suggestion
	// when multiple trailing zero bytes, format as N * 2**x
```
I think that distinction is handled by `sign` and only relevant at the `return`, so no need to have a multiple line comment.

Ruko97 replied:
I wanted to write `format as ±N * 2**x` in the comment here to indicate that the printed variable is shown with its sign, but I wasn't sure if Unicode is supported in comments.

matheusaaguiar replied:
Right, fair enough. I've noticed that also that you changed the templates to `N * 2**x - 1` and `-N * 2**x + 1` , so it seems worth having a comment about this.

matheusaaguiar reviewed (changes requested):
Sorry, I had another look and found some things that were glossed over before and could be improved.

leonardoalt commented:
Auto please rebase onto origin/develop to get rid of the CI errors.

Ruko97 commented:
I have rebased and updated my commit based on the suggestions here. Sorry for being late.

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,111 @@ std::string joinHumanReadablePrefixed
+/// Returns decimal representation for smaller numbers; hex for large numbers.
+/// "Special" numbers, powers-of-two and powers-of-two minus 1, are returned in
+/// formulaic form like 0x01 * 2**24 - 1.
+/// @a T can be any integer value, will typically be u160, u256 or bigint.
```
I don't want to be pedantic but it looks to me that `T` refers to a type like `u160` or `bigint`. It is the function param variable   `_value` that stores the integer value of type `T`.
```suggestion
/// @a T can be any integer type, will typically be u160, u256 or bigint.
```

Ruko97 replied:
Sorry, I accidentally might have forgot to add this edit to my code. My bad. It's fixed now.

matheusaaguiar replied:
No problem :)

matheusaaguiar commented:
> I have rebased and updated my commit based on the suggestions here. Sorry for being late.

No need to apologize, you are not late at all. You are actually quite timely with your responses. Thanks for contributing to the project :)

matheusaaguiar reviewed (commented):
I consider it ready. But it needs another rebase on top of `develop` since we are having a release today.

matheusaaguiar reviewed (dismissed):
It is going to need another rebase, but I am approving so someone else from the team can merge it later.

nikola-matic commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,109 @@ std::string joinHumanReadablePrefixed
-			" * 2**" +
-			std::to_string(i * 8);
+		if (temp == 1)
+			return sign + "2**" + std::to_string(i * 8);
```
Hey, sorry for being a little late to the party. Can you please swap all instances of string concatenation with `fmt::format`. 
E.g., in this case, it would be something like this:
```
return fmt::format("{}2**{}", sign, std::to_string(i * 8);
```
Otherwise, looks good!

Ruko97 replied:
ok, I am doing it now.

nikola-matic commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,109 @@ std::string joinHumanReadablePrefixed
+				std::to_string(i * 8);
 	}
 
 	// when multiple trailing FF bytes, format as N * 2**x - 1
```
Shouldn't this now be commented as `2**x  - 1` instead of `N * 2**x  - 1`? Or at least that `N` is optional depending on whether the number is a power of 2?

Ruko97 replied:
No, not always. `N` can be unequal to 1 if the variable is not exactly a power of 2 or one-less-than a power of 2. For example, `9 * 2**249` would be displayed as `18 * 2**248`.

Ruko97 replied:
Perhaps I should add a test regarding this.

matheusaaguiar replied:
That is a good idea. We can never have enough tests.

matheusaaguiar commented on test/libsolutil/StringUtils.cpp:
```diff
@@ -190,7 +206,7 @@ BOOST_AUTO_TEST_CASE(test_format_number_readable_signed)
 	BOOST_CHECK_EQUAL(formatNumberReadable(b, true), "-0x5555...{+56 more}...5555");
 	BOOST_CHECK_EQUAL(formatNumberReadable(c, true), "-0x0BCD...{+56 more}...6789");
-	BOOST_CHECK_EQUAL(formatNumberReadable(d, true), "-0x5555555555555556 * 2**192 - 1");
+	BOOST_CHECK_EQUAL(formatNumberReadable(d, true), "-0x5555555555555556 * 2**192 + 1");
```
Does this change has to do with the bug you mentioned, as in
> `formatNumberReadable(-0xffffffff)` should be equal to `-2 ** 32 + 1` but previously it was tested to be equal to `-2 ** 32 - 1`

?

Also, it is a bit difficult to track what these hex values translate to as decimals. How are you making sure that this is indeed the right value?

Ruko97 replied:
Yes, the change is related to the bug I mentioned
Much of the test values is carried from previous values that were being tested. I only changed some of the variables to negative. I also used `isoltest` to verify the values.

axic commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,109 @@ std::string joinHumanReadablePrefixed
 	int i = 0;
-	T v = _value;
-	for (; (v & 0xff) == 0; v >>= 8)
+	for (temp = signedValue; (temp & 0xff) == 0; temp >>= 8)
```
I think it may be better scoping `temp` here and in the next loop, as they are not used inbetween.

axic replied:
Would also keep it named `v` as before.

Ruko97 replied:
wouldn't using `temp` here make it more clear what the role of the variable is? It took me some time to understand what role `v` played in the code; and I think other programmers would face a similar situation when looking at the code.

matheusaaguiar replied:
I agree with naming the variable as `temp` and also with limiting its scope to only inside the loops.

Ruko97 replied:
ok, will do that

axic commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,109 @@ std::string joinHumanReadablePrefixed
 
 	HexCase hexcase = HexCase::Mixed;
-	HexPrefix prefix = HexPrefix::Add;
+	HexPrefix hexprefix = HexPrefix::Add;
```
I'd also name these `hexCase` and `hexPrefix`.

axic commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +101,109 @@ std::string joinHumanReadablePrefixed
+
+	bool isNegative = _value < 0;
+	bigint temp;
+	bigint signedValue = isNegative ? (bigint(-1) * _value) : bigint(_value);
```
Since this is copied, perhaps make it const. And `sign` too.

Ruko97 commented:
Has `fmt::format` shifted to some other file? The errors in CircleCI seem to be related to not being able to find `fmt::format`.

cameel commented:
It's there but we added it as a dependency very recently so not all of our libs link to it yet. You have to add it to `target_link_libraries` in `libsolutil/CMakeLists.txt`. See how #13473 does it.

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
+		{
+			// 0x100 yields 2**8 (N is 1 and redundant)
+			if (temp == 1)
+				return fmt::format("{}2**{}", sign, std::to_string(i * 8));
```
```suggestion
				return fmt::format("{}2**{}", sign, i * 8);
```
I guess we can remove the conversion to string now, since `fmt::format` should be able to handle it.

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
+				int j = 0;
+				for (; (temp & 0x1) == 0; temp >>= 1)
+					j++;
+				return fmt::format("{}2**{}", sign, std::to_string(i * 8 + j));
```
```suggestion
				return fmt::format("{}2**{}", sign, i * 8 + j);
```

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
+					"{}{} * 2**{}",
+					sign,
+					toHex(toCompactBigEndian(temp), hexPrefix, hexCase),
+					std::to_string(i * 8)
```
```suggestion
					i * 8
```

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
+			std::string suffix = isNegative ? " + 1" : " - 1";
+
+			if (temp == 0)
+				return fmt::format("{}2**{}{}", sign, std::to_string(i * 8), suffix);
```
```suggestion
				return fmt::format("{}2**{}{}", sign, i * 8, suffix);
```

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
+				int j = 0;
+				for (; (temp & 0x1) != 0; temp >>= 1)
+					j++;
+				return fmt::format("{}2**{}{}", sign, std::to_string(i * 8 + j), suffix);
```
```suggestion
				return fmt::format("{}2**{}{}", sign, i * 8 + j, suffix);
```

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
+					"{}{} * 2**{}{}",
+					sign,
+					toHex(toCompactBigEndian(++temp), hexPrefix, hexCase),
+					std::to_string(i * 8),
```
```suggestion
					i * 8,
```

matheusaaguiar replied:
This is still pending.

Ruko97 replied:
fixed

matheusaaguiar commented on libsolutil/StringUtils.h:
```diff
@@ -101,109 +103,119 @@ std::string joinHumanReadablePrefixed
+			"{}{}...{{+{} more}}...{}",
+			sign,
+			str.substr(0, initialChars),
+			std::to_string(numSkipped),
```
```suggestion
			numSkipped,
```

matheusaaguiar replied:
This is still pending.

Ruko97 replied:
fixed

matheusaaguiar reviewed (commented):
Some adjustments and I wanted to confirm that you added the test case mentioned [here](https://github.com/ethereum/solidity/pull/13277#discussion_r978648348).
Also, @nikola-matic are you satisfied with the current replacements of string concatenation or do you think that they are needed for the simple cases at lines `133`, `203` and `218`?

nikola-matic commented:
> Some adjustments and I wanted to confirm that you added the test case mentioned [here](https://github.com/ethereum/solidity/pull/13277#discussion_r978648348). Also, @nikola-matic are you satisfied with the current replacements of string concatenation or do you think that they are needed for the simple cases at lines `133`, `203` and `218`?

The formatting looks good now, and no, no need for `fmt::format` for simple cases. Another rebase would be nice to get rid of the `gp2` failure.

Ruko97 commented:
> Some adjustments and I wanted to confirm that you added the test case mentioned [here](https://github.com/ethereum/solidity/pull/13277#discussion_r978648348). Also, @nikola-matic are you satisfied with the current replacements of string concatenation or do you think that they are needed for the simple cases at lines `133`, `203` and `218`?

Yes, I added them at lines `185`, `186`, `285` and `286`. I added some more tests now just to make sure.

nikola-matic commented:
@matheusaaguiar @ekpyron will this need a changelog entry?

ekpyron commented:
> @matheusaaguiar @ekpyron will this need a changelog entry?

I'd say, it's borderline. But strictly speaking it's user-facing, so we can just default to "yes".

matheusaaguiar commented:
@Ruko97 , please add a [Changelog](https://github.com/ethereum/solidity/blob/46a7ebd5b66c5638427295b9c3dbf15fc48c4ed2/Changelog.md) entry under the `Compiler Features`, describing shortly the improvement on readability of large/hex numbers.

nikola-matic commented on Changelog.md:
```diff
@@ -8,9 +8,11 @@ Compiler Features:
 * Standard JSON: Add a boolean field `settings.metadata.appendCBOR` that skips CBOR metadata from getting appended at the end of the bytecode.
 * Yul Optimizer: Allow replacing the previously hard-coded cleanup sequence by specifying custom steps after a colon delimiter (``:``) in the sequence string.
 * Language Server: Add basic document hover support.
+* SMTChecker: Improved readability for large integers that are powers of two or almost powers of two
```
@leonardoalt is this SMT checker related?

nikola-matic commented:
Rebased, resolved conflicts, and added periods to changelog entires. The build should now be fine, and hopefully closer to merging.

NunoFilipeSantos reviewed (approved):
LGTM

matheusaaguiar reviewed (approved):
LGTM
